rust/src/lib.rs

```diff
@@ -0,0 +1,262 @@
+        let encrypted = self.encrypt(data).await.map_err(|e| e.to_string())?;
+        self.put_s3_obj(
+            ByteStream::from(encrypted.aes_gcm_ciphertext),
+            &format!("{}.aesgcm.encrypted", name),
+        )
+        .await
+        .map_err(|e| e.to_string())?;
+
+        self.put_s3_obj(
+            ByteStream::from(encrypted.data_key),
+            &format!("{}.key", name),
+        )
+        .await
+        .map_err(|e| e.to_string())?;
+
+        self.put_s3_obj(
+            ByteStream::from(encrypted.meta.as_bytes().to_owned()),
+            &format!("{}.meta", name),
+        )
+        .await
+        .map_err(|e| e.to_string())?;
+
+        Ok(())
+    }
+
+    pub async fn lookup(&self, name: &str) -> Result<String, String> {
+        let key = name;
+        let data_key = self.get_s3_obj_as_vec(format!("{key}.key")).await.unwrap();
+        let meta_add = self.get_s3_obj_as_vec(format!("{key}.meta")).await.unwrap();
+        let ciphertext = self
+            .get_s3_obj_as_vec(format!("{key}.aesgcm.encrypted"))
+            .await?;
+        let meta: Meta = serde_json::from_slice(&meta_add).map_err(|e| e.to_string())?;
+        let cipher: AesGcm<Aes256, typenum::U12> =
+            AesGcm::new_from_slice(self.direct_decrypt(&data_key).await?.as_slice()).unwrap();
+        let nonce = general_purpose::STANDARD
+            .decode(meta.nonce)
+            .map_err(|e| e.to_string())?;
+        let nonce = Nonce::from_slice(nonce.as_slice());
+        let res = cipher
+            .decrypt(
+                nonce,
+                Payload {
+                    msg: &ciphertext,
+                    aad: &meta_add,
+                },
+            )
+            .map_err(|e| e.to_string())?;
+        Ok(String::from_utf8(res).unwrap())
+    }
+}
+
+struct EncryptObject {
+    data_key: Vec<u8>,
+    aes_gcm_ciphertext: Vec<u8>,
+    meta: String,
+}
+
+fn parse_output_value_from_key(key: &str, out: &[Output]) -> Option<String> {
+    out.iter()
+        .find(|output| output.output_key() == Some(key))
+        .map(|output| output.output_value().unwrap_or_default().to_owned())
+}
+
+async fn get_cf_params(config: &SdkConfig, stack: &str) -> Result<CfParams, String> {
+    let stack_output = cfClient::new(config)
+        .describe_stacks()
+        .stack_name(stack)
+        .send()
+        .await
+        .map_err(|err| err.to_string())?
+        .stacks()
+        .and_then(|stacks| stacks.first())
+        .and_then(|stack| stack.outputs())
+        .ok_or("error getting Cloudformation Stack Ouput")?
+        .to_owned();
+
+    Ok(CfParams {
+        bucket_name: parse_output_value_from_key("vaultBucketName", &stack_output)
+            .expect("Error getting bucket name from stack"),
+        key_arn: parse_output_value_from_key("kmsKeyArn", &stack_output),
+        // deployed_version: parse_output_value_from_key("vaultStackVersion", &stack_output),
```

Comment:
Commented out, should be removed?

Reply:
If in the future we want move completely to a Rust-based CLI, this is needed for updating the vault Cloudformation stack. This can be removed now and added later.